### DISCUSSION TITLE
Async atoms are not memoized

### DISCUSSION THREAD

**Discussion by chrison94** [pmndrs/jotai#1925] [General]
Given an async atom if we return two times the same promise, downstream atoms are notified with a different promise. 

That happens because a promise value is not checked against the previous atom value.  That behaviour is from my perspective unexpected. Was this intended?

Relevant Code - https://github.com/pmndrs/jotai/blob/main/src/vanilla/store.ts#L227-L287

If necessary, I can provide a reproduction.

**Comment by dai-shi** [comment]
Yeah, currently we always create a cancellable promise and wrap with it. So, the behavior is intentional.
But, I wish it would be better.
Yeah, can you create a very small example with https://csb.jotai.org ?

**Reply by chrison94** [reply]
Sync example with automatic memoizing - https://codesandbox.io/s/exciting-austin-i8pg0w
Async example without memoizing -  https://codesandbox.io/s/practical-night-qpnd4j?file=/src/App.tsx

I think it would be enough if we only wrap the cancellable promise if the new promise is indeed a different one than before

**Reply by backbone87** [reply]
Any plan on how to proceed with this? I can provide failing test cases and a draft PR for a potential fix

**Reply by dai-shi** [reply]
I can take time for this in a couple of days, but if you can draft a PR, it would be helpful, especially for failing tests.

**Reply by backbone87** [reply]
created a PR with failing test case and fix
